[docs] Add EAS Tutorial

amandeepmittal opened the pull request:
# Why

<!--
Please describe the motivation for this PR, and link to relevant GitHub issues, forums posts, or feature requests.
-->

Add EAS tutorial section in Learn.

# Test Plan

See preview: /tutorial/eas/introduction/

<!--
Please describe how you tested this change and how a reviewer could reproduce your test, especially if this PR does not include automated tests! If possible, please also provide terminal output and/or screenshots demonstrating your test/reproduction.
-->

# Checklist

<!--
Please check the appropriate items below if they apply to your diff. This is required for changes to Expo modules.
-->

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

amandeepmittal committed:
Create EAS tutorial section in learn & add first two chapters

amandeepmittal committed:
Fix chapter URLs

amandeepmittal committed:
Chapter 1 initial draft

amandeepmittal committed:
Chapter 2 initial draft

amandeepmittal committed:
Chapter 3 initial draft

amandeepmittal committed:
Chapter 3 initial draft

amandeepmittal committed:
Chapter 4 initial draft

amandeepmittal committed:
Chapter 5 initial draft

amandeepmittal committed:
Chapter 7 initial draft

amandeepmittal committed:
Chapter 8 initial draft

amandeepmittal committed:
Chapter 9 initial draft

amandeepmittal committed:
Chapter 10 initial draft

amandeepmittal committed:
Fix title, slug, and description

amandeepmittal committed:
Apply feedback from Jon

amandeepmittal committed:
Fix images not rendering for iOS device

amandeepmittal committed:
Fix minor issues

amandeepmittal committed:
Fix minor issues, update verbiage, & github draft

amandeepmittal committed:
Add code highlights

amandeepmittal committed:
Add internal links

amandeepmittal committed:
Add Next steps page

amandeepmittal committed:
Fix vale warnings

kadikraman commented on docs/pages/tutorial/eas/android-development-build.mdx:
```diff
@@ -0,0 +1,191 @@
+
+This section provides the methods available for running the development build on an Android device:
+
+- Open with Expo Orbit (macOS only)
```
Windows and Linux support [coming soon](https://linear.app/expo/project/orbit-v200-support-on-windows-and-linux-03a10a450949) 😄

amandeepmittal replied:
When that lands, we'll update this tutorial and hopefully make the next step as an alternate.

amandeepmittal replied:
I think I'm going to remove "macOS only" part as I heard from Brent the support for Windows is coming soon.

kadikraman commented on docs/pages/tutorial/eas/ios-development-build-for-devices.mdx:
```diff
@@ -0,0 +1,229 @@
+
+Development builds for iOS devices are generated in the **.ipa** format, which is standard for iOS app installations.
+
+> **Note**: Apple Developer membership is required to create and install a development build on an iOS device.
```
Can we put the simulator build instructions before the device build instructions? I think a lot of people will hold out on getting an Apple Developer account if they can so they'd want to skip this and only so the simulator build.

kadikraman replied:
I wonder if in general it's worth explaining build signing before diving into the registration? Especially the why you need an apple dev account.

amandeepmittal replied:
> Can we put the simulator build instructions before the device build instructions? I think a lot of people will hold out on getting an Apple Developer account if they can so they'd want to skip this and only so the simulator build.

That's a good point! I'll adjust the navigation accordingly to put the iOS simulator before iOS in the navigation structure.

amandeepmittal replied:
> I wonder if in general it's worth explaining build signing before diving into the registration? Especially the why you need an apple dev account.

I'm currently thinking of providing this context as: "Each build is signed with [necessary credentials](/app-signing/app-credentials/#ios) to ensure that the app is from a trusted source. **Apple Developer account membership is required because it provides access to these credentials for signing our app.** EAS Build simplifies this process by managing those credentials for us."

- It provides link to learn more about what those credentials are in the App credentials doc.
- Do you think it answers the question? Or is it still complicated?

kadikraman replied:
I think that's perfect: enough info to provide context, but not a massive detour from the lesson point.

kadikraman commented on docs/pages/tutorial/eas/ios-development-build-for-devices.mdx:
```diff
@@ -0,0 +1,229 @@
+
+<Step label="2">
+
+## Download and install the provisioning profile
```
Hm, you know I've never used that flow before so I can't actually comment on it. I keep it simple by registering the device with the UDID and adding it to the profile. Maybe we can provide that as an alternative?

amandeepmittal replied:
"I keep it simple by registering the device with the UDID and adding it to the profile"

- Can you elaborate on this more or give me tips so that I can try it myself? I don't think I've used this method.

kadikraman replied:
Haha, maybe it's best to keep yours then!

But to explain, UDID is basically this (you can see it in Finder when you plug your phone in). It's a unique ID for your device. You copy it, and if you choose the UDID option it asks you to name the device and add that ID. Then you add that device to the Ad Hoc provisioning profile and that's it.


<img width="1032" alt="Screenshot 2024-02-21 at 20 15 35" src="https://github.com/expo/expo/assets/6534400/f778b440-e704-4b7b-b43b-83e371c9e37b">

amandeepmittal replied:
Thanks for sharing these details! I think you are right about adding this as an alternative. I'll try this on my end and just to be able to document the steps in this process. Will follow up on this separately (maybe as a separate task a bit later).

kadikraman commented on docs/pages/tutorial/eas/ios-development-build-for-devices.mdx:
```diff
@@ -0,0 +1,229 @@
+
+<Step label="3">
+
+## Create a development build
```
Same as for Android, it would be handy to refer back to the part of `eas.json` we're talking about.

kadikraman commented on docs/pages/tutorial/eas/ios-development-build-for-devices.mdx:
```diff
@@ -0,0 +1,229 @@
+
+> **info** **Tip**: Next time you run `eas build` command, you can also use `-p` to specify the platform. It is short for `--platform`.
+
+This command prompts us with the following questions:
```
It only does this the first time you build.

kadikraman commented on docs/pages/tutorial/eas/ios-development-build-for-simulators.mdx:
```diff
@@ -0,0 +1,153 @@
+
+<Terminal cmd={['$ eas build --platform ios --profile ios-simulator']} />
+
+This command prompts us with the following question:
```
It only does this the first time you build.

kadikraman commented on docs/pages/tutorial/eas/internal-distribution-builds.mdx:
```diff
@@ -0,0 +1,191 @@
+import { BoxLink } from '~/ui/components/BoxLink';
+import { BookOpen02Icon } from '@expo/styleguide-icons';
+
+In this chapter, we're going to learn how to set up [internal distribution builds](/guides/sharing-preview-releases/#internal-distribution). Creating an internal build is quite similar to creating a development build. However, the difference is how it is used and distributed.
```
```suggestion
In this chapter, we're going to learn how to set up [internal distribution builds](/guides/sharing-preview-releases/#internal-distribution). Creating an internal build is quite similar to creating a development build. The difference is that you don't need to run the development server locally for it to work.
```

kadikraman commented on docs/pages/tutorial/eas/internal-distribution-builds.mdx:
```diff
@@ -0,0 +1,191 @@
+
+## Internal distribution build
+
+An internal distribution build is great for sharing an app's progress with our team (technical and non-technical stakeholders) to gather feedback without running a development server. Our team can directly install the build and run it like an app on their device.
```
Our -> Your?

amandeepmittal replied:
For the tutorial (this one and the previous one), we are continuing the tradition of using first-person voice. I know it might be confusing to go against the writing style guide but I didn't want to create parity between the two tutorials 😄

kadikraman commented on docs/pages/tutorial/eas/internal-distribution-builds.mdx:
```diff
@@ -0,0 +1,191 @@
+An internal distribution build is great for sharing an app's progress with our team (technical and non-technical stakeholders) to gather feedback without running a development server. Our team can directly install the build and run it like an app on their device.
+
+## Ways to distribute an app internally
+
```
Seems like there should be a sentence here about how both Apple and Google have built-in mechanisms for sharing an app internally etc.

kadikraman commented on docs/pages/tutorial/eas/internal-distribution-builds.mdx:
```diff
@@ -0,0 +1,191 @@
+
+This is all we need to create our first internal distribution build. The `preview` build profile from the above snippet has a `distribution` property whose value is set to `internal`. This value allows us to share our build URLs with anyone so they to install it on their device and do not require a development server to run the app.
+
+As discussed in the previous chapters, for non-app store builds, Android requires **.apk** and iOS needs **.ipa** formats. This applies to internal distribution builds as well. The `distribution` when set to `internal`, automatically creates the app binary in these file formats.
```
"The distribution when set to internal, automatically creates the app binary in these file formats." -> only for device builds. On a simulator it builds an `app` as you said.

amandeepmittal replied:
Add the context, thank you for catching this!

kadikraman commented on docs/pages/tutorial/eas/manage-app-versions.mdx:
```diff
@@ -0,0 +1,86 @@
+
+In this chapter, we're going to configure our app to auto-increment the developer-facing app version. Learning about it will be useful before we dive into production build in the next few chapters.
+
+## Difference between developer-facing and user-facing app versions
```
Should we also explain why there is a `version` _and_ a build number and the rules around them?

- once you've uploaded an app with a version and build number to apple / google, you can't use the same version & build number combination again
- on android, the build number must always be increasing (so e.g. if you've used 5, you can't use 4 even if you never created a build with 4)
- the idea is that as you're working on e.g. 1.0.0 of your app and need to upload new versions, you increment the build number only until that version is published. Once you publish 1.0.0 you can't upload another build with this version so you have to increment it

amandeepmittal replied:
I've added explanation in context to app stores and relying on the same terminology to help explain things a bit better.

kadikraman commented on docs/pages/tutorial/eas/manage-app-versions.mdx:
```diff
@@ -0,0 +1,86 @@
+  "cli": {
+    /* @hide ... */ /* @end */
+    /* @info Add the appVersionSource property and set it to remote*/
+    "appVersionSource": "remote"
```
Do we want to recommend this? It makes me uncomfortable not managing it manually, it's not clear what makes it increment.

kadikraman commented on docs/pages/tutorial/eas/android-production-build.mdx:
```diff
@@ -0,0 +1,318 @@
+
+## Add Google Service Account permissions key
+
+From now on, we can use [EAS Submit](/submit/introduction/) to automate releases and avoid the manual process. To do that, we need to add the service account key to our project's **eas.json**.
```
You can only do this if you fill out _all_ the metadata (including app assets etc) and it also has to go through a review process. Maybe this should be in a separate lesson?

amandeepmittal replied:
We've a callout and a tip about this before this section starts (which I think will help users to pay attention to.):

![CleanShot 2024-05-07 at 01 03 05](https://github.com/expo/expo/assets/10234615/c2f91389-e6dc-4395-aa95-cc769598b7df)

kadikraman commented on docs/pages/tutorial/eas/android-production-build.mdx:
```diff
@@ -0,0 +1,318 @@
+To publish and distribute an app on the Google Play Store, we need to:
+
+- Have a paid developer account. To create one, follow the instructions on the [Google Play sign-up page](https://play.google.com/apps/publish/signup/).
+- Google Service Account email and JSON key. To create one, follow the instructions from our [expo.fyi](https://github.com/expo/fyi/blob/main/creating-google-service-account.md) guide. Once you've followed the steps, come back to this guide.
```
This should be optional as it's only required if you want to submit automatically

amandeepmittal replied:
I think this our recommended path so it's might be confusing for new users as to identify what's the recommended approach here will be for them. Introducing conditions might lead to a bit more confusion.

kadikraman commented on docs/pages/tutorial/eas/team-development.mdx:
```diff
@@ -0,0 +1,184 @@
+    },
+    "ios-simulator": {
+      /* @hide ... */ /* @end */
+      "channel": "ios-simulator"
```
Shouldn't the channel be "development" here?

kadikraman replied:
I know you _can_, I just meany why would you have a separate channel for the iOS simulator, surely you'd just use the `development` channel.

amandeepmittal replied:
When there is an additional build profile in **eas.json** (apart from the default ones) and on running `eas update:configure`, it creates a channel for each profile based on the its name automatically. In a way, this is the default configuration based on the steps followed in the tutorial. I didn't originally thought of changing it. However, I see your point from app's development point of view. For example, my app is in development phase, I want to keep the development related config same for all environments.

I think I'll just update this in the tutorial and make a note for the reader to remember removing it before proceeding since the `ios-simulator` profile already extends `development` profile.

amandeepmittal replied:
Added the callout describing the above behaviour.

kadikraman commented on docs/pages/tutorial/eas/using-github.mdx:
````diff
@@ -0,0 +1,151 @@
+}
+```
+
+- Next, modify our app's JavaScript code. We create a git branch called `dev` and submit a PR to our example repository from that branch.
````
```suggestion
- Next, create a new branch called `dev`, make a change in your app's JavaScript code, commit it, push the branch and create a PR from that branch.
```

amandeepmittal replied:
Thanks for the suggestion! I slightly modified it to keep the second person voice intact here (hope you won't mind).

kadikraman reviewed (commented):
Wow, that's a huge amount of work, congrats! 👏
I added some comments based on my first read-though.

Simek commented on docs/pages/tutorial/eas/using-github.mdx:
````diff
@@ -0,0 +1,151 @@
+
+- We need to specify the build image that we will be using. Open **eas.json**, and under the `development` profile, add [`android.image`](/eas/json/#image) and [`ios.image`](/eas/json/#image-1) properties and set their value to [`latest`](/build-reference/infrastructure/#configuring-build-environment).
+
+```json eas.json
+{
+  "build": {
+    "development": {
+      /* @hide ... */ /* @end */
+      /* @info */
+      "android": {
+        "image": "latest"
+      },
+      "ios": {
+        "image": "latest"
+      }
+      /* @end */
+    }
+  }
+  /* @hide ... */ /* @end */
+}
+```
````
```suggestion
  ```json eas.json
  {
    "build": {
      "development": {
        /* @hide ... */ /* @end */
        /* @info */
        "android": {
          "image": "latest"
        },
        "ios": {
          "image": "latest"
        }
        /* @end */
      }
    }
    /* @hide ... */ /* @end */
  }
  ```
```

Formatting nit, this should indent code block to match list bullet.

amandeepmittal committed:
Apply suggestions from code review

Co-authored-by: Kadi Kraman <kadi@expo.io>
Co-authored-by: Bartosz Kaszubowski <gosimek@gmail.com>

amandeepmittal commented on docs/pages/tutorial/eas/introduction.mdx:
```diff
@@ -0,0 +1,57 @@
+
+- Use the Sticker Smash app from the previous tutorial. If you haven't followed the previous tutorial, download the complete source code from [this GitHub repository](https://github.com/amandeepmittal/sticker-smash-expo-tutorial-app) and set the project on your local machine. **We're going to use this as our example app**.
+- Create a new project with [`npx create-expo-app`](/get-started/create-a-project/).
+- Create and use a project generated by React Native CLI (also known as bare React Native project). It will contain the native directories (**android** and **ios**). You'll have to install the `expo` package (either [automatically](/bare/installing-expo-modules/) or [manually](/bare/installing-expo-modules/#manual-installation)) in the project before you proceed.
```
Do you mean changing the title of the section to Prerequisites? Or create a sub section here for it?

kadikraman replied:
I think I originally meant having a section called "Prerequisites", but looking back again, just renaming the section would work as well, because it tells you, in order to do this tutorial, you need a React Native project: either use an existing project, create one or use Sticker Smash

amandeepmittal commented on docs/pages/tutorial/eas/android-development-build.mdx:
```diff
@@ -0,0 +1,191 @@
+
+To create a **.apk**, make sure:
+
+- The `developmentClient` is set to `true` in **eas.json**, which is the default configuration.
```
Yeah, I did mean to explain that it is set by default on development profile and you are right that this property "by default" is false (as it is not set for other profiles). Thanks for catching this!

amandeepmittal commented on docs/pages/tutorial/eas/android-development-build.mdx:
```diff
@@ -0,0 +1,191 @@
+
+For Android devices or emulators, the development build must be in the [**.apk** file format](/build-reference/apk/). While the default Android format is **.aab**, which is ideal for Google Play Store distribution, it cannot be directly installed on devices or emulators.
+
+To create a **.apk**, make sure:
```
Great catch! adding that context here.

amandeepmittal committed:
Apply feedback from Kadi

amandeepmittal commented on docs/pages/tutorial/eas/configure-development-build.mdx:
```diff
@@ -0,0 +1,217 @@
+
+## Benefits of creating a development build
+
+A development build offers several advantages. Here are some key highlights:
```
I'm updating with just a table. I think if we add the table, there is a little requirement for this list since there is an overlap.

amandeepmittal committed:
Apply feedback from Kadi

quinlanj reviewed (approved):
🚀

amandeepmittal committed:
Apply suggestions from code review

Co-authored-by: Quinlan Jung <quin@expo.io>

amandeepmittal committed:
Apply feedback from Brent's review

amandeepmittal committed:
Apply feedback from code review

amandeepmittal committed:
Apply feedback from code review

amandeepmittal committed:
Change callout type to a info

brentvatne commented on docs/pages/tutorial/eas/introduction.mdx:
```diff
@@ -0,0 +1,58 @@
+import { BoxLink } from '~/ui/components/BoxLink';
+import { BookOpen02Icon } from '@expo/styleguide-icons';
+
+In this tutorial, we'll take our app built in the [previous tutorial](/tutorial/introduction/), create a development build and a preview version, and deploy it to the app stores.
```
```suggestion
In this tutorial, we'll take our app built in the [previous tutorial](/tutorial/introduction/), create a development build and a preview version, and deploy it to app stores.
```

(and also update other places where we say "the app stores" to just "app stores" - there are many app stores and google play / apple are just the biggest ones)

brentvatne commented on docs/pages/tutorial/eas/introduction.mdx:
```diff
@@ -0,0 +1,58 @@
+
+## About this tutorial
+
+The objective of this tutorial is to give you basic proficiency [Expo Application Services (EAS)](https://expo.dev/eas) core services: [Build](/build/introduction/), [Submit](/submit/introduction/), and [Update](/eas-update/introduction/). The following topics are covered:
```
```suggestion
The objective of this tutorial is to give you basic proficiency with [Expo Application Services (EAS)](https://expo.dev/eas) core services: [Build](/build/introduction/), [Submit](/submit/introduction/), and [Update](/eas-update/introduction/). When you complete the tutorial, you will know how to set up a professional mobile Android and iOS development and deployment CI/CD pipeline for your individual and team projects. The following topics are covered:
```

brentvatne commented on docs/pages/tutorial/eas/introduction.mdx:
```diff
@@ -0,0 +1,58 @@
+The objective of this tutorial is to give you basic proficiency [Expo Application Services (EAS)](https://expo.dev/eas) core services: [Build](/build/introduction/), [Submit](/submit/introduction/), and [Update](/eas-update/introduction/). The following topics are covered:
+
+- Use EAS Build to create and install a development build, then run it on a device, emulator, or simulator.
+- Discover the benefits of using a development build instead of Expo Go.
```
```suggestion
- Experience the benefits of using a development build instead of Expo Go.
```

brentvatne commented on docs/pages/tutorial/eas/configure-development-build.mdx:
```diff
@@ -0,0 +1,220 @@
+| **Collaboration**                 | Facilitates team testing with shared native runtime.                                                                                           | Easy project sharing via QR codes on a device.                                                       |
+| **Third-party libraries support** | Full support for any [third-party library](/workflow/using-libraries/#third-party-libraries), including those that require custom native code. | Limited to libraries within the Expo SDK, not suitable for custom native dependencies.               |
+| **Customization**                 | Extensive customization with [config plugins](/config-plugins/introduction/) and direct access to native code.                                 | Limited customization with a focus on Expo SDK capabilities without direct native code modification. |
+| **App size**                      | Optimized app size, including only the libraries used in the project.                                                                          | Generally larger due to the inclusion of the entire Expo SDK.                                        |
```
```suggestion
```

brentvatne replied:
this isn't really useful information. app size doesn't matter for development builds. in fact, your development build might even be larger than expo go distributed through the app store/play store because it will bit split for your specific device when installed through the store

amandeepmittal committed:
Apply suggestions from code review

Co-authored-by: Brent Vatne <brentvatne@gmail.com>

amandeepmittal committed:
Change "the app stores" to "app stores"

amandeepmittal committed:
Update link to Review > Overview

amandeepmittal committed:
Update screenshot of the new "Preview" button

amandeepmittal committed:
Update next-steps.mdx

amandeepmittal committed:
[docs] Add a progress tracker for EAS Tutorial (#27971)

# Why

<!--
Please describe the motivation for this PR, and link to relevant GitHub
issues, forums posts, or feature requests.
-->

Fixes ENG-11468

Add a Progress Tracker component to track progress when a chapter is
completed in the EAS Tutorial. It:
- shows a success message when the user reaches the end of the chapter
with the summary of that chapter
- allows the user to mark a chapter read/unread
- syncs the state of chapter (read/unread) with the sidebar
- adds and updates the sidebar circular progress on reaching tutorial
complition
- allows user to reset the progress on reaching tutorial complition

Also, handles Google Analytics 4 event to send a custom event when all
chapters in a tutorial are completed. The event added is called
`eas_tutorial`.

# Test Plan

To test it, see the **Summary** section of a chapter. For example:
http://docs.expo.dev-pr-27971.s3-website-us-east-1.amazonaws.com/tutorial/eas/configure-development-build/#summary.

## Preview

### Saving Progress Tracker for a chapter



https://github.com/expo/expo/assets/10234615/ced38210-e4ea-4c6e-b7d9-578f1a4ee5db



### Completed Progress Tracker and reset tutorial functionality



https://github.com/expo/expo/assets/10234615/3d671c8b-6f79-4c5c-a166-b88a03f6afbd






<!--
Please describe how you tested this change and how a reviewer could
reproduce your test, especially if this PR does not include automated
tests! If possible, please also provide terminal output and/or
screenshots demonstrating your test/reproduction.
-->


# Checklist

<!--
Please check the appropriate items below if they apply to your diff.
This is required for changes to Expo modules.
-->

- [ ] Documentation is up to date to reflect these changes (eg:
https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style
Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build
(eg: updated a module plugin).

amandeepmittal committed:
Merge branch 'main' into aman/eng-11036-create-a-pr-for-eas-tutorial

amandeepmittal committed:
Fix prettier lint issues

amandeepmittal committed:
Apply feedback part 1

amandeepmittal committed:
Update

amandeepmittal committed:
Update GitHub guide

amandeepmittal committed:
Apply feedback from review

amandeepmittal committed:
Update descriptions

amandeepmittal committed:
Add Learn > Overview section

amandeepmittal committed:
Update overview.mdx

brentvatne reviewed (approved):
![image](https://github.com/expo/expo/assets/90494/6808ac85-15db-422e-a3ff-1d5696d3074b)

let's ship this and we can iterate on it! thank you @amandeepmittal!

amandeepmittal committed:
revert some changes

amandeepmittal committed:
Remove 'Done' and show progress when all chapters completed

amandeepmittal commented on docs/pages/tutorial/eas/using-github.mdx:
```diff
@@ -0,0 +1,134 @@
+
+[Expo GitHub App](/build/building-from-github/) automatically triggers builds from our GitHub projects with EAS. We can trigger builds for any build profile based on our development team's preferences. It also allows triggering builds for `git` push committed directly to a repository or a pull request.
+
+In this chapter, we're to configure this functionality. We already have a GitHub repository for our example app to demonstrate this.
```
```suggestion
In this chapter, we'll configure this functionality. We already have a GitHub repository for our example app to demonstrate this.
```

amandeepmittal committed:
Fix typo

amandeepmittal committed:
Apply feedback from Simek's review

Simek reviewed (approved):
Looks great! 🔥 Thanks for all the work put into creating this tutorial! 🙏

amandeepmittal committed:
Minor fixes
